### PR TITLE
Add reachability checks for cover properties

### LINKED ISSUE #2023: Distinguish between unsatisfiable and unreachable cover statements
For the following program, Kani currently reports the first `cover` statement as `UNSATISFIABLE`. It would be more informative if it reported it as `UNREACHABLE` to indicate that the statement itself is unreachable, and not just that the condition cannot be satisfied.

```rust
#[kani::proof]
fn main() {
    let x = if kani::any() { 2 } else { 3 };
    if x > 5 {
        kani::cover!(x == 6); // unreachable
    } else {
        kani::cover!(x == 2);
    }
}
```

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Check the reachability of `cover` properties, and report unreachable ones as such.

### Resolved issues:

Resolves #2023 

### Related RFC:

Optional #ISSUE-NUMBER.

### Testing:

* How is this change tested? Added a new test

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs [hunk 1]
```diff
@@ -144,6 +145,33 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    ) -> (String, Stmt) {
+        if self.queries.get_check_assertion_reachability() {
+            // Generate a unique ID for the assert
+            let assert_id = self.next_check_id();
+            // Generate a message for the reachability check that includes the unique ID
+            let reach_msg = GotocCtx::reachability_check_message(&assert_id);
+            // Also add the unique ID as a prefix to the assert message so that it can be
+            // easily paired with the reachability check
+            let msg = GotocCtx::add_prefix_to_msg(&msg, &assert_id);
+            // inject a reachability (cover) check
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs:168]: I think calling this a cover is now a bit confusing. Shouldn't we start using a different property class for reachability checks and cover for actual cover? Or am I missing something?

**Review by celinval — COMMENTED** [review]
The PR itself is OK, but I was wondering if we should have a different property class for reachability checks. I think it is a bit confusing to have a cover to cover a cover. :grin:

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs (hunk 1 above)
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs:168]: Good point. I'll change their property class to something else.

Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs]: Done: I introduced a new property class, `PropertyClass::ReachabilityCheck`, and switched the CBMC output parser to identify reachability checks using the property class (as opposed to the description as was done previously).

Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs]: Thanks!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs [hunk 2]
```diff
@@ -134,9 +137,41 @@ impl<'tcx> GotocCtx<'tcx> {
     // The above represent the basic operations we can perform w.r.t. assert/assume/cover
     // Below are various helper functions for constructing the above more easily.
 
-    /// A shorthand for cover(true)
-    pub fn codegen_cover_loc(&self, msg: &str, span: Option<Span>) -> Stmt {
-        self.codegen_cover(Expr::bool_true(), msg, span)
+    /// Given the message for an assertion, generate a reachability check that
+    /// is meant to check whether the assertion is reachable. The function
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs:141]: s/assertion/property

File: kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs [hunk 3]
```diff
@@ -228,14 +227,8 @@ impl<'tcx> GotocCtx<'tcx> {
 
                 // TODO: switch to tagging assertions via the property class once CBMC allows that:
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs:228]: Should we remove this TODO?

File: kani-driver/src/cbmc_property_renderer.rs [hunk 4]
```diff
@@ -622,25 +624,11 @@ fn remove_check_ids_from_description(mut properties: Vec<Property>) -> Vec<Prope
...
-        if prop.description.contains(message) {
-            removed_properties.push(prop);
-        } else {
-            filtered_properties.push(prop);
-        }
-    }
-    (filtered_properties, removed_properties)
-}
-
 /// Filters reachability checks with `filter_properties`
```
Comment (celinval) [inline kani-driver/src/cbmc_property_renderer.rs:627]: update comment

File: tests/expected/cover/cover-unreachable/expected [hunk 5]
```diff
@@ -0,0 +1,16 @@
+Status: UNREACHABLE\
+Description: "cover condition: x == 2"\
+main.rs:11:13 in function cover_unreachable
```
Comment (celinval) [inline tests/expected/cover/cover-unreachable/expected:3]: please remove line numbers. :)

**Review by celinval — APPROVED** [review]
Just some minor comments. Thanks!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs [hunk 6]
```diff
@@ -59,6 +60,8 @@ pub enum PropertyClass {
     ///
     /// SPECIAL BEHAVIOR: None TODO: Why should this exist?
     FiniteCheck,
+    /// Checks added by Kani compiler to determine whether an assertion is reachable
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs]: Done

File: kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs (hunk 2 above)
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs]: Done

File: kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs (hunk 3 above)
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs]: Yes. Removed.

File: kani-driver/src/cbmc_property_renderer.rs (hunk 4 above)
Comment (zhassan-aws) [reply kani-driver/src/cbmc_property_renderer.rs]: Done

File: tests/expected/cover/cover-unreachable/expected (hunk 5 above)
Comment (zhassan-aws) [reply tests/expected/cover/cover-unreachable/expected]: Done

**Pull request merged**